Eclipse Maven Install fails to replace built  .war file, even with -SNAPSHOT version

I have a Maven Java project in Eclipse, and I select the project, then "Run As->Maven install"

This runs, and "packages the webapp", "assembles the webapp","processes the war project","builds the war",

When it comes to "spring-boot-maven-plugin:2.5.4:repackage (repackage)
[INFO] Replacing main artifact with repackaged archive" it fails.

[ERROR] Failed to execute goal com.ibm.cics:cics-bundle-maven-plugin:1.0.2:bundle-war (default) on project  Error writing bundle resource

If I re-run with -X I find out it is because it already exists.

This is apparently normal Maven behaviour, as it should not replace the same version. I can re-version (but that makes no sense for a development project) OR I can add -SNAPSHOT to the end of the version.

I add -SNAPSHOT to the end of the version and it works... because it creates a new file.

I do it again... and it fails because it already exists.

If I do "maven clean install" it works. Shouldn't -SNAPSHOT replace this file?

Sure. Re-running now.

[INFO] Scanning for projects...

[INFO] 

[INFO] ---------< com.ibm.cics.cip.bank.springboot:customerservices >----------

[INFO] Building customerservices 1.1.0-SNAPSHOT

[INFO] --------------------------------[ war ]---------------------------------

[INFO] 

[INFO] --- maven-resources-plugin:3.2.0:resources (default-resources) @ customerservices ---

[INFO] Using 'UTF-8' encoding to copy filtered resources.

[INFO] Using 'UTF-8' encoding to copy filtered properties files.

[INFO] Copying 1 resource

[INFO] Copying 12 resources

[INFO] 

[INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ customerservices ---

[INFO] Nothing to compile - all classes are up to date

[INFO] 

[INFO] --- maven-resources-plugin:3.2.0:testResources (default-testResources) @ customerservices ---

[INFO] Using 'UTF-8' encoding to copy filtered resources.

[INFO] Using 'UTF-8' encoding to copy filtered properties files.

[INFO] skip non existing resourceDirectory *redactedLocalDirectory*\src\test\resources

[INFO] 

[INFO] --- maven-compiler-plugin:3.8.1:testCompile (default-testCompile) @ customerservices ---

[INFO] No sources to compile

[INFO] 

[INFO] --- maven-surefire-plugin:2.22.2:test (default-test) @ customerservices ---

[INFO] No tests to run.

[INFO] 

[INFO] --- maven-war-plugin:3.3.1:war (default-war) @ customerservices ---

[INFO] Packaging webapp

[INFO] Assembling webapp [customerservices] in [*redactedLocalDirectory*\target\customerservices-1.1.0-SNAPSHOT]

[INFO] Processing war project

[INFO] Building war: *redactedLocalDirectory*\target\customerservices-1.1.0-SNAPSHOT.war

[INFO] 

[INFO] --- spring-boot-maven-plugin:2.5.4:repackage (repackage) @ customerservices ---

[INFO] Replacing main artifact with repackaged archive

[INFO] 

[INFO] --- cics-bundle-maven-plugin:1.0.2:bundle-war (default) @ customerservices ---

[INFO] ------------------------------------------------------------------------

[INFO] BUILD FAILURE

[INFO] ------------------------------------------------------------------------

[INFO] Total time: 27.804 s

[INFO] Finished at: 2022-03-28T11:22:51+01:00

[INFO] ------------------------------------------------------------------------

[ERROR] Failed to execute goal com.ibm.cics:cics-bundle-maven-plugin:1.0.2:bundle-war (default) on project customerservices: Error writing bundle resource "customerservices-1.1.0-SNAPSHOT.warbundle": *redactedLocalDirectory*\target\customerservices-1.1.0-SNAPSHOT-cics-bundle\customerservices-1.1.0-SNAPSHOT.warbundle -> [Help 1]

[ERROR] 

[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.

[ERROR] Re-run Maven using the -X switch to enable full debug logging.

[ERROR] 

[ERROR] For more information about the errors and possible solutions, please read the following articles:

[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException

--- a/cics-bundle-maven-plugin/src/it/test-repeat-bundle-war/invoker.properties
+++ b/cics-bundle-maven-plugin/src/it/test-repeat-bundle-war/invoker.properties
@@ -1,0 +1,17 @@
+###
+# #%L
+# CICS Bundle Maven Plugin
+# %%
+# Copyright (C) 2022 IBM Corp.
+# %%
+# This program and the accompanying materials are made
+# available under the terms of the Eclipse Public License 2.0
+# which is available at https://www.eclipse.org/legal/epl-2.0/
+# 
+# SPDX-License-Identifier: EPL-2.0
+# #L%
+###
+# We run the build more than once by using the xxx.N syntax for any property.
+# We don't actually want to change the properties, just run it twice.
+invoker.profiles.1=
+invoker.profiles.2=

--- a/cics-bundle-maven-plugin/src/it/test-repeat-bundle-war/pom.xml
+++ b/cics-bundle-maven-plugin/src/it/test-repeat-bundle-war/pom.xml
@@ -1,0 +1,71 @@
+<!--
+  #%L
+  CICS Bundle Maven Plugin
+  %%
+  Copyright (C) 2022 IBM Corp.
+  %%
+  This program and the accompanying materials are made
+  available under the terms of the Eclipse Public License 2.0
+  which is available at https://www.eclipse.org/legal/epl-2.0/
+  
+  SPDX-License-Identifier: EPL-2.0
+  #L%
+  -->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>com.ibm.cics.test-bundle-war</groupId>
+  <artifactId>test-repeat-bundle-war</artifactId>
+  <version>0.0.1-SNAPSHOT</version>
+  <packaging>war</packaging>
+  
+  <dependencies>
+    <!-- provided -->
+    <dependency>
+      <groupId>javax.ws.rs</groupId>
+      <artifactId>javax.ws.rs-api</artifactId>
+      <version>2.0</version>
+      <scope>provided</scope>
+    </dependency>
+  </dependencies>
+  
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-war-plugin</artifactId>
+        <version>3.2.3</version>
+        <configuration>
+          <failOnMissingWebXml>false</failOnMissingWebXml>
+          <packagingExcludes>pom.xml</packagingExcludes>
+          <warName>test-war</warName>
+        </configuration>
+      </plugin>
+      
+      <plugin>
+        <artifactId>maven-compiler-plugin</artifactId>
+        <version>3.5.1</version>
+        <configuration>
+          <source>1.8</source>
+          <target>1.8</target>
+        </configuration>
+      </plugin>
+      
+      <plugin>
+        <groupId>@project.groupId@</groupId>
+        <artifactId>@project.artifactId@</artifactId>
+        <version>@project.version@</version>
+        <executions>
+          <execution>
+            <goals>
+              <goal>bundle-war</goal>
+            </goals>
+            <configuration>
+              <jvmserver>EYUCMCIJ</jvmserver>
+            </configuration>
+          </execution>
+        </executions>
+      </plugin>
+    </plugins>
+  </build>
+  
+</project>

--- a/cics-bundle-maven-plugin/src/it/test-repeat-bundle-war/src/main/java/test_war/TestEndpoint.java
+++ b/cics-bundle-maven-plugin/src/it/test-repeat-bundle-war/src/main/java/test_war/TestEndpoint.java
@@ -1,0 +1,39 @@
+package test_war;
+
+/*-
+ * #%L
+ * CICS Bundle Maven Plugin
+ * %%
+ * Copyright (C) 2022 IBM Corp.
+ * %%
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ * #L%
+ */
+
+import java.util.Optional;
+
+import javax.ws.rs.ApplicationPath;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+@ApplicationPath("")
+@Path("")
+public class TestEndpoint extends javax.ws.rs.core.Application {
+	
+	public TestEndpoint() {
+	}
+
+	@GET
+	@Produces(MediaType.APPLICATION_OCTET_STREAM)
+	public Response get() {
+		return Response.ok().build();
+	}
+	
+}

--- a/cics-bundle-maven-plugin/src/main/java/com/ibm/cics/cbmp/AbstractBundleJavaMojo.java
+++ b/cics-bundle-maven-plugin/src/main/java/com/ibm/cics/cbmp/AbstractBundleJavaMojo.java
@@ -15,8 +15,10 @@ package com.ibm.cics.cbmp;
  */
 
 import java.io.File;
+import java.io.IOException;
 import java.nio.file.Path;
 
+import org.apache.commons.io.FileUtils;
 import org.apache.maven.artifact.Artifact;
 import org.apache.maven.plugin.MojoExecutionException;
 import org.apache.maven.plugin.MojoFailureException;
@@ -64,6 +66,18 @@ public abstract class AbstractBundleJavaMojo extends AbstractBundlePublisherMojo
 
 	@Override
 	public void execute() throws MojoExecutionException, MojoFailureException {
+		
+		if (workDir.exists()) {
+			getLog().debug("Deleting " + workDir);
+			try {
+				FileUtils.deleteDirectory(workDir);
+			} catch (IOException e) {
+				throw new MojoExecutionException("Unable to delete CICS bundle output directory " + workDir, e);
+			}
+		}
+		
+		workDir.mkdirs();
+		
 		BundlePublisher bundlePublisher = getBundlePublisher();
 		
 		Artifact artifact = project.getArtifact();

